### ISSUE TITLE
WorkContextAccessor shouldn't use ThreadStatic for work contexts created independent from an http context

### ISSUE THREAD

**Issue by orchardbot** [OrchardCMS/Orchard#4338]
@Piedone created:
https://orchard.codeplex.com/workitem/20509

If there is no http context present then WorkContextAccessor stores created work contexts in a thread-static dictionary. The issue with this implementation is that it relies on the same thread running a given execution path from start to finish (what is generally an issue in ASP.NET). This is certainly not the case however if the execution path contains async code since then the execution can start with one thread and (after resuming from an async call) continue with another one.

Instead WorkContextAccessor could use System.Runtime.Remoting.Messaging.CallContext to store the work contexts, similarly to how the Orchard.Environment.State.ContextState<T> object works. CallContext.LogicalSet/GetData() could work for this but it probably isn't practically possible since it would need the WorkContext to be immutable.

So I'm not sure how this can be solved. Probably WorkContextAccessor could enable the use of an external context object (apart from HttpContext) that is maintained by the caller, and thus can be carried in an arbitrary way? A hackish workaround available now is to explicitly set an http context through IHttpContextAccessor before creating the WC and then using that to carry the work contexts through layers.

Related: https://orchard.codeplex.com/workitem/20474

**Comment by Piedone** [comment]
What have you done regarding this @pszmyd ?

**Comment by jtkech** [comment]
@Piedone, @pszmyd, Not yet comfortable with the concepts involved here, so **be indulgent because I'm sure that I'm wrong and it would be great for me to know why**.

Here, some tests that work on a local environment with 2 tenants, but with a simple installation with no specific async code...

Without removing [ThreadStatic], why use a concurrent dictionary if there is one dictionary per thread. So, 1st test done with a normal dictionary where, under debugger, I could see and confirm that there was always one workcontext per dictionary (not tested, but if only one workcontext, why use a dictionary). Of course, this doesn't solve the async issue.

Or why use [ThreadStatic] if we have a concurrent dictionary. In this 2nd test I could see sometimes the 2 tenant work contexts in the dictionary. Only without [ThreadStatic], maybe this could work with async code.

But maybe a ConcurrentDictionary isn't totally thread safe and / or it's not good to have many tenant work contexts in the same dictionary because of its not so good internal locking system...

Maybe we would need, not a [ThreadStatic] attribute, but something like a "[TenantStatic]" attribute. But, isn't it what Autofac do for us when we register WorkContextAccessor with an InstancePerMatchingLifetimeScope("shell")?

So, in the last test, I've removed, not only the [ThreadStatic] attribute, but also all the static keywords related to this dictionary (here a concurrent dictionary because without [ThreadStatic]). This because I said to me that Autofac will create a different but only one WorkContextAccessor instance per shell / tenant, so it will be the same for its non static members...

In this last test, I could see that there was always one workcontext per dictionary. But here there are multiple dictionaries not because of different threads, but because of different shell scopes. Because without [ThreadStatic], maybe this could work with async code.

But, in this last case, even with only one workcontext per dictionary, maybe some drawbacks with the thread safing of a concurrent dictionary when async code. As I remember, with a concurrent dictionary, a potential problem is when 2 threads create an entry in the same time (e.g with AddOrUpdate(), because of the use of a delegate). But if not the same shell / tenant it would not be the same dictionary, and if the same tenant with async code, I'm not sure this scenario is possible (e.g if async code is triggered when the work context scope is already created).

Note: During a web request, I thought before that we could have nested work context scopes. But, after reading the code, I can see that there is only one entry in the httpContext.Items that is removed when a child scope is disposed... Indeed, with nested scopes, I get some errors in the parent scope when trying to use the work context after a child scope is disposed...

Best

**Comment by Piedone** [comment]
Yes, a ConcurrentDictionary wouldn't be a good solution here, I too think.

You can't rely on Autofac here, as the WorkContextScope should somehow be externally "carried": you need something in IWorkContextAccessor.GetContext() that will magically retrieve the "current" WC.

Now in a request this is easy since there's the HttpContext and that carries the WCS: that's why you didn't see any use of the static field during a request. And this means that a request will be still fine, even if it uses async (and is thus possibly served by multiple threads).

The issue comes when there is no HttpContext (e.g. in a background task): then you need something else to carry the WC, and here comes the static field into play. However, since it's ThreadStatic it won't work for async code (which can, and most possibly will, switch threads during execution).

**Comment by sebastienros** [comment]
So the issue only comes with asyn/await that we are not using, so I could say there are no issues here. Should we close the bug?

FYI there are new CallContext.LogicalGetData / CallContext.LogicalSetData methods, working with immutable data. We could try that instead of the ThreadStatic, they work with async/await, but again, no issues ;)

**Comment by jtkech** [comment]
@Piedone thanks a lot for your clarifications

**Comment by jtkech** [comment]
@Piedone Thanks again for your comments that confirmed me some things... and where the main point is when you said:
> **You can't rely on Autofac here**...

So, I've done some tests with `2 tenants` and in an `IBackgroundTask.Sweep()` method I've run in parallel `2 multi threads tasks`. And the results seems to confirm what I was saying (see my previous comments). But maybe again I missed something else, sorry for my stupidity.

I've used the same following code in 3 contexts (see below for the descriptions and results). Here the code in a `Sweep()` method where I uses a WCA belonging to the parent scope but where the `GetContext() method is called in different threads`.

            var testTasks = new List<Task>();
            for (var i = 0; i < 2; i++) {
                var testTask = Task.Factory.StartNew(() => {
                    var workContext = _wca.GetContext();
                    var culture = ""; var siteName = "";

                    lock (_synLock) {
                        culture = workContext.CurrentCulture;
                        siteName = workContext.CurrentSite.SiteName;
                    }

                    var threadId = System.Threading.Thread.CurrentThread.ManagedThreadId;
                    Logger.Error("SiteName: \"{0}\" ThreadId: \"{1}\" Culture: \"{2}\".",
                        siteName, threadId, culture);
                });

                testTasks.Add(testTask);
            }
            Task.WaitAll(testTasks.ToArray());

1. With a `[ThreadStatic] static ConcurrentDictionary`, as currently, I got a lot of errors because one dictionary per thread...

2. Just removed the `[ThreadStatic]` attribute, so the dictionary becomes a pure static one. So, there is one dictionary for all threads but each tenant uses a different entry key created with the tenant instance of the `WorkContextAccessor`. This seems to work but maybe not good to have all tenant workcontexts in the same dictionary, e.g because of some non thread safe methods of a `ConcurrentDictionary` (see my previous post).

3. I've removed not only the `[ThreadStatic]` attribute, but also the `static keyword` of the dictionary
    `//[ThreadStatic]`
    `/*static*/ ConcurrentDictionary<object, WorkContext> _threadStaticContexts;`
    `/*static*/ ConcurrentDictionary<object, WorkContext> EnsureThreadStaticContexts() {`
Here, the dictionary isn't static but, with the `WorkContextAccessor`, an instance is created for each tenant. It's like a `[TenantStatic]` attribute, then there is one dictionary per tenant but that can be accessed by different threads. And this seems to work.

When it works, each time the background task is executed, the following is added to the log file

    ....SiteName: "Orchard Dev" ThreadId: "40" Culture: "en-US". [(null)]
    ....SiteName: "Orchard Dev" ThreadId: "50" Culture: "en-US". [(null)]
    ....SiteName: "Alpha" ThreadId: "15" Culture: "fr-FR". [(null)]
    ....SiteName: "Alpha" ThreadId: "43" Culture: "fr-FR". [(null)]

**This confirms that each tenant uses 2 different threads and get the right workcontext used to acquire the site name and current culture.**

> **Note about some concurrency issues in my tests**

The workcontext can be got by different threads in the same time, but not some of its properties. But here I think it's because of other non thread safe implementations. Indeed, only for testing, if I use a `_synLock object in the WorkContextImplementation GetState() method`, it works. Another workaround was to do a first synchronous access (in a first task that run alone, or in the parent scope). Finally, for my tests, a `_synLock object that belongs to the parent scope is used in each task`.

> **Note about async code with only the async and await keywords**

I could apply the `async` keyword to the Sweep() method itself, and then I've used the `await` keyword on other `async methods`... What I've seen is that there is no new threads created, so, if you don't start this programming model in a new thread, it seems to work with the current implementation.

Sorry for the length of this post

Best

**Comment by Piedone** [comment]
@sebastienros "we" are not using async as in there's no async background code in Orchard but I think it's not too hard to see why you'd want to write such in your own application. I mentioned CallContext.LogicalSet/GetData() in the opening post but they're not applicable here.

@jtkech now that you say it I also have a hard time understanding why the dictionary is thread-static at all. However, your test has a flaw: it operates with the assumption that there will be just one work context for all the threads at a given time. This is actually not what should happen: if you write multi-threaded code in Orchard you should use a different work context for each thread, since IDependency implementation are most possibly not thread-safe.

So to summarize, we have the following aim, for background (non-HttpContext-accessing) tasks:

Async code should work, the current work context should be carried during execution:
```
_wca.GetContext(); // This should return the same WC as below.
await DoAsync(); // Imagine that after this line execution continues on a different thread.
_wca.GetContext(); // This should return the same WC as above.
```

Furthermore if we go into the nuances, this should also work, again in a bg task:
```
        var testTasks = new List<Task>();
        for (var i = 0; i < 2; i++) {
            var testTask = Task.Factory.StartNew(() => {
                using (var workContext= _wca.CreateWorkContextScope()) {
					var culture = ""; var siteName = "";

					// Async inside a thread! Wow.
					_wca.GetContext(); // This should return the same WC as below.
					await DoAsync(); // Imagine that after this line execution continues on a different thread.
					_wca.GetContext(); // This should return the same WC as above.

					lock (_synLock) {
						culture = workContext.CurrentCulture;
						siteName = workContext.CurrentSite.SiteName;
					}

					var threadId = System.Threading.Thread.CurrentThread.ManagedThreadId;
					Logger.Error("SiteName: \"{0}\" ThreadId: \"{1}\" Culture: \"{2}\".",
						siteName, threadId, culture);
				}
            });

            testTasks.Add(testTask);
        }
        Task.WaitAll(testTasks.ToArray());
```

While the second example is obviously complicated (probably too much) but it demonstrates that in the end it needs the same infrastructure in place than the first one.

**Comment by jtkech** [comment]
@Piedone Oh, yes I understand if `we should use a different work context for each thread` because of `IDependency implementation are most possibly not thread-safe`

So, I have to read more here: http://docs.autofac.org/en/latest/advanced/concurrency.html#thread-safe-types where they are saying for example:

`So, a correct Autofac application will use a ContainerBuilder on a single thread to create the container at startup. Subsequent use of the container can occur on any thread.`

About the `Async inside a thread! Wow.` where I had to add the async keyword to the lambda expression, not sure, but as I've seen for now, the line after an `await keyword` is executed in the same thread. As it seems to be said in this doc: https://msdn.microsoft.com/fr-fr/library/vstudio/Hh191443(v=VS.110).aspx where they are saying for example:

`The async and await keywords don't cause additional threads to be created.`

Best

**Comment by jtkech** [comment]
`

@Piedone, now I totally understand what you mean. In my first tests, the `await` was done on an async call that didn't use any `Task` method that could switch to a new thread. Now, I can do it by simply using `await Task.Delay(n)`...

So, what do you think of this very simple new `CallContextScopeImplementation class` (see below) that seems to work in my tests with 2 tenants * 2 parallel tasks with some `await`, workcontext set / get, services resolving... Each parallel task creates a new scope which saves the workcontext in the `Callcontext`. This forces a copy-on-write with only shallow copies of data, but now the `Callcontext` references a new workcontext resolved from a new lifetime scope, so, `immutabibility` doesn't seem to be a problem.

In the rest of the execution path, around the await keywords, the CallContext is carried among thread switches. And here also, I don't think `immutabibility` is a problem, because here it's in a kind of "linear" execution flow where only one thread (that uses the same workcontext scope) runs at a time.


        class CallContextScopeImplementation : IWorkContextScope {
            readonly WorkContext _workContext;
            readonly Action _disposer;

            public CallContextScopeImplementation(IEnumerable<IWorkContextEvents> events,
                ILifetimeScope lifetimeScope) {
                _workContext = lifetimeScope.Resolve<WorkContext>();
                CallContext.LogicalSetData("WorkContext", _workContext);

                _disposer = () => {
                    events.Invoke(e => e.Finished(), NullLogger.Instance);
                    CallContext.FreeNamedDataSlot("WorkContext");
                    lifetimeScope.Dispose();
                };
            }
        // rest of the class
        ...
        }

Usage and 2 modified methods for testing

        // usage by passing a null httpContext parameter
        using (_wca.CreateWorkContextScope(null)) {
            ...
        }

        // modified method with null checking on the httpContext parameter
        public IWorkContextScope CreateWorkContextScope(HttpContextBase httpContext) {
            var workLifetime = _lifetimeScope.BeginLifetimeScope("work");
            if (httpContext != null)
                workLifetime.Resolve<WorkContextProperty<HttpContextBase>>().Value = httpContext;
            ...
            if (httpContext == null)
                return new CallContextScopeImplementation(events, workLifetime);
            ...
        }

        // modified method to get the context from the CallContext
        public WorkContext GetContext() {
            ...
            if (CallContext.LogicalGetData("WorkContext") != null)
                return CallContext.LogicalGetData("WorkContext") as WorkContext;
            ...
        }

Best

**Comment by Piedone** [comment]
@jtkech Yes, "The async and await keywords don't cause additional threads to be created." but I wasn't implying this. Async-await however can (and most possibly will) cause a *thread switch*: e.g. your code executes on thread 1, then asynchronously waits for some I/O task to complete, then gets to picked up by thread 2 (basically another, not necessarily the same thread from the thread pool; the original thread may be doing something completely different now).

Your implementation with CallContext is interesting and would love to see it indeed flawlessly work. BTW what happens if you modify the WC (e.g. write CurrentUser)? Is this properly reflected? Could you do some more tests where this is there instead of the ThreadStatic implementation for background tasks?

**Comment by jtkech** [comment]
In my tests, in SweepGenerator.cs I use CreateWorkContextScope(**null**), so it's the CallContext which is used. Then in a IBackgroundTask I have 2 parallel tasks with their own scope created in the same way. So we have nested scopes based on the CallContext.

Then, in each parallel task I use some await, services resolving... And in one of the task (executed twice because of 2 tenants) I modify the culture before an await that is correctly retrieved after the await even it's not in the same thread.

So, not tested with the CurrentUser, but it works with the CurrentCulture.

**Comment by jtkech** [comment]
If the test above on the CurrentCulture above isn't enough.
Please give me a piece of code that write the CurrentUser, it's not allowed to do that directly.
Then I could test it.

Best

**Comment by jtkech** [comment]
Test done with the current user with this code in a parallel task:

    var user = scope.Resolve<IMembershipService>().GetUser("jtk");
    scope.WorkContext.SetState<IUser>("CurrentUser", user);
    ...
    await Task.Delay(150);
    ....
    var userName = scope.WorkContext.CurrentUser != null ?
        scope.WorkContext.CurrentUser.UserName : "None";
    ....

After the `await Task.Delay(150)`, even the continuation code is on another thread, this is properly reflected. But not on another task that runs in parallel in its own context scope.

**Comment by jtkech** [comment]
@Piedone, yes, I had seen that after an await you can continue in another thread, that's why I use now `await Task.Delay(n)`, before I was using `await Thread.Sleep(n)...`

I've tried your suggestion without the `ThreadStaticScopeImplementation`, only the `CallContextScopeImplementation`. Each parallel task has its own workcontext overwritten in the `CallContext`, and the workcontext is carried among a thread switch after an `await`...

All seems to work fine.

But I'm aware that this needs to be seriously evaluated and maybe not practicable in some environments, but I have already learned a lot... So, just for playing, in the context of my tests described above, I could add this in `BackgroundService`.

    public void Sweep() {
        var asyncTasks = new List<Task>(); 
        foreach (var task in _tasks) {
            var asyncTask = Task.Run(() => {
                using (var scope = _wca.CreateWorkContextScope()) {
                    ...
                    var transactionManager = scope.Resolve<ITransactionManager>();
                    try {
                        ...
                        transactionManager.RequireNew();
                        task.Sweep();
                        ...
                    }
                    catch (Exception e) {
                        transactionManager.Cancel();
                        ...
                    }
                }
            });
            asyncTasks.Add(asyncTask);
        }
        Task.WaitAll(asyncTasks.ToArray());
    }

Best

**Comment by jtkech** [comment]
Note: I've updated my previous comment where I call background tasks in parallel tasks. I needed to resolve each IBackgroundTask in its own context scope...

@Piedone Here the complete code of the `CallContextScopeImplementation` class in `WorkContextAccessor.cs` that works (less lines than the original file) but in a lightweight local environment.

    using System;
    using System.Collections.Concurrent;
    using System.Collections.Generic;
    using System.Runtime.Remoting.Messaging;
    using System.Web;
    using Autofac;
    using Orchard.Logging;
    using Orchard.Mvc;
    using Orchard.Mvc.Extensions;

    namespace Orchard.Environment {
      public class WorkContextAccessor : IWorkContextAccessor {
        readonly ILifetimeScope _lifetimeScope;

        readonly IHttpContextAccessor _httpContextAccessor;
        // a different symbolic key is used for each tenant.
        // this guarantees the correct accessor is being resolved.
        readonly object _workContextKey = new object();

        public WorkContextAccessor(
            IHttpContextAccessor httpContextAccessor,
            ILifetimeScope lifetimeScope) {
            _httpContextAccessor = httpContextAccessor;
            _lifetimeScope = lifetimeScope;
        }

        public WorkContext GetContext(HttpContextBase httpContext) {
            // because can be called directly
            if (!httpContext.IsBackgroundContext())
                return httpContext.Items[_workContextKey] as WorkContext;

            return CallContext.LogicalGetData("WorkContext") as WorkContext;
        }

        public WorkContext GetContext() {
            var httpContext = _httpContextAccessor.Current();
            if (!httpContext.IsBackgroundContext())
                return GetContext(httpContext);

            return CallContext.LogicalGetData("WorkContext") as WorkContext;
        }

        public IWorkContextScope CreateWorkContextScope(HttpContextBase httpContext) {
            var workLifetime = _lifetimeScope.BeginLifetimeScope("work");
            if (!httpContext.IsBackgroundContext())
                workLifetime.Resolve<WorkContextProperty<HttpContextBase>>().Value = httpContext;

            var events = workLifetime.Resolve<IEnumerable<IWorkContextEvents>>();
            events.Invoke(e => e.Started(), NullLogger.Instance);

            if (httpContext.IsBackgroundContext())
                return new CallContextScopeImplementation(events, workLifetime);

            return new HttpContextScopeImplementation(
                events,
                workLifetime,
                httpContext,
                _workContextKey);
        }

        public IWorkContextScope CreateWorkContextScope() {
            var httpContext = _httpContextAccessor.Current();
            if (!httpContext.IsBackgroundContext())
                return CreateWorkContextScope(httpContext);

            var workLifetime = _lifetimeScope.BeginLifetimeScope("work");

            var events = workLifetime.Resolve<IEnumerable<IWorkContextEvents>>();
            events.Invoke(e => e.Started(), NullLogger.Instance);

            return new CallContextScopeImplementation(events, workLifetime);
        }

        class HttpContextScopeImplementation : IWorkContextScope {
            readonly WorkContext _workContext;
            readonly Action _disposer;

            public HttpContextScopeImplementation(IEnumerable<IWorkContextEvents> events, ILifetimeScope lifetimeScope, HttpContextBase httpContext, object workContextKey) {
                _workContext = lifetimeScope.Resolve<WorkContext>();
                httpContext.Items[workContextKey] = _workContext;

                _disposer = () => {
                    events.Invoke(e => e.Finished(), NullLogger.Instance);

                    httpContext.Items.Remove(workContextKey);
                    lifetimeScope.Dispose();
                };
            }

            void IDisposable.Dispose() {
                _disposer();
            }

            public WorkContext WorkContext {
                get { return _workContext; }
            }

            public TService Resolve<TService>() {
                return WorkContext.Resolve<TService>();
            }

            public bool TryResolve<TService>(out TService service) {
                return WorkContext.TryResolve(out service);
            }
        }

        class CallContextScopeImplementation : IWorkContextScope {
            readonly WorkContext _workContext;
            readonly Action _disposer;

            public CallContextScopeImplementation(IEnumerable<IWorkContextEvents> events, ILifetimeScope lifetimeScope) {
                _workContext = lifetimeScope.Resolve<WorkContext>();
                CallContext.LogicalSetData("WorkContext", _workContext);

                _disposer = () => {
                    events.Invoke(e => e.Finished(), NullLogger.Instance);
                    CallContext.FreeNamedDataSlot("WorkContext");
                    lifetimeScope.Dispose();
                };
            }

            void IDisposable.Dispose() {
                _disposer();
            }

            public WorkContext WorkContext {
                get { return _workContext; }
            }

            public TService Resolve<TService>() {
                return WorkContext.Resolve<TService>();
            }

            public bool TryResolve<TService>(out TService service) {
                return WorkContext.TryResolve(out service);
            }
        }
      }
    }

Best

**Comment by jtkech** [comment]
@Piedone just to say that I've done an experimental PR #5966 to share my code and for testing. In this PR, I also carry the httpcontext (e.g in background tasks) to be used by the HttpContextAccessor, and I've done others adaptations related to other issues and requirements...

**Comment by Piedone** [comment]
Thank you @jtkech .

**Comment by Piedone** [comment]
Was fixed in https://github.com/OrchardCMS/Orchard/pull/5966.
